Added contrbutors OG image using genratemetadata

dgparmar14 opened the pull request:
# Description

Pull request for feature enhancements of contributor's OG image showcasing their merged PRs, reviewed PRs, and feed updates, akin to the highlight section. Additionally, it incorporates the contributor's GitHub image, GitHub username, and description as per their profile. The Open-Graph image also includes the organization logo.

fixes #127 

### Here are some examples of how it looks

![Screenshot 2024-03-20 141803](https://github.com/coronasafe/leaderboard/assets/99136186/94f25dc6-7e86-409e-b6ef-58056e2754b7)
![Screenshot 2024-03-20 141743](https://github.com/coronasafe/leaderboard/assets/99136186/badc6786-80d3-477c-8024-58cd011b73cf)

dgparmar14 commented:
@rithviknishad Please review this pr and let's close the issue

rithviknishad commented on public/logo.png:
Organization specific data is not kept in this repo. Could you move this to the data repo like how we have the `logo.webp`? Also name this as `logo-black.png` instead. Or instead, you could store the color logo itself and instead use tailwind classes to make the image render as black using invert and brightness classes

rithviknishad commented on app/api/og/route.tsx:
Use tailwindcss instead of css

rithviknishad commented on app/api/og/Roboto-Bold.ttf:
Use google fonts instead

rithviknishad reviewed (changes requested):
Accidentally clicked approve for previous review.... :/

dgparmar14 commented:
> https://vercel.com/docs/functions/og-image-generation/og-image-examples#using-tailwind-css---experimental

Sure I will do it,
I used that but I dropp the idea because of its still experimental,

dgparmar14 commented:
@rithviknishad, your requested changes have been implemented:

1. Changed `logo.webp` to `logo.jpg` and created a `symlink` for this.
2. Moved Roboto-Bold.ttf to the public directory since Google Font is not supported at the edge time.
3. Converted all possible CSS to Tailwind CSS in the OG route. However, please note that `Tailwind CSS` is experimental in og-images, so some styling remains in normal CSS.
4. Added metadataBase: new URL(env.NEXT_PUBLIC_META_URL) in the `layout.tsx` file to address the mentioned error.

![Screenshot 2024-03-23 102106](https://github.com/coronasafe/leaderboard/assets/99136186/f48811e9-6c59-405e-a482-4e29f3dfff55)

Let me know if there's anything else you need!

rithviknishad commented on app/api/og/route.tsx:
```diff
@@ -0,0 +1,92 @@
+  const profileImage = `https://avatars.githubusercontent.com/${github_username}`;
+
+  const fontData = await fetch(
+    new URL("../../../public/Roboto-Bold.ttf", import.meta.url),
```
contents in public directory can be accessed directly like this:
```suggestion
    new URL("/Roboto-Bold.ttf", import.meta.url),
```

<img width="1600" alt="image" src="https://github.com/coronasafe/leaderboard/assets/25143503/9aee74f2-9839-446d-9ad5-8ddae9bec545">

rithviknishad reviewed (changes requested):
More such issues present... Try doing a local build and preview


<img width="1516" alt="image" src="https://github.com/coronasafe/leaderboard/assets/25143503/97528f7f-8282-4021-9fd8-4665d6dccc8b">

rithviknishad commented:
@dgparmar14 I assume whats happening is it is getting pre-rendered and so those files wouldn't be served until the build is completed and output is served right?

Have you tried importing the font from `next/fonts` instead? We could also disable pre-renders for this maybe.

dgparmar14 commented:
> @dgparmar14 I assume whats happening is it is getting pre-rendered and so those files wouldn't be served until the build is completed and output is served right?
> 
> Have you tried importing the font from `next/fonts` instead? We could also disable pre-renders for this maybe.

Yes I tried but it doesn't work at the edge time
And above error is because of first need to merge pr of data repo.

dgparmar14 commented:
> > @dgparmar14 I assume whats happening is it is getting pre-rendered and so those files wouldn't be served until the build is completed and output is served right?
> > 
> > Have you tried importing the font from `next/fonts` instead? We could also disable pre-renders for this maybe.
> 
> Yes I tried but it doesn't work at the edge time
> And above error is because of first need to merge pr of data repo.
> and i also run build locally and it is run perfectly fine with my changes.

rithviknishad commented:
Reviewed the data-repo PR. Also it'd be great if you could mention related PR's from data-repo in this PR's body itself. It's nice to keep track of related PRs documented in the PR body.

rithviknishad commented:
> and i also run build locally and it is run perfectly fine with my changes

Are you sure it's working locally? Not sure if I'm doing something wrong... I did checkout the data-repo to the one you've made and tried running this branch with it and getting the following errors upon visiting the profile of a contributor. Also getting 404 page not found for all contributors.

<img width="3120" alt="image" src="https://github.com/coronasafe/leaderboard/assets/25143503/fd96d75d-18d2-4a16-b026-5234e8bef9ec">

dgparmar14 commented:
> > and i also run build locally and it is run perfectly fine with my changes
> 
> Are you sure it's working locally? Not sure if I'm doing something wrong... I did checkout the data-repo to the one you've made and tried running this branch with it and getting the following errors upon visiting the profile of a contributor. Also getting 404 page not found for all contributors.
> 
> <img width="3120" alt="image" src="https://github.com/coronasafe/leaderboard/assets/25143503/fd96d75d-18d2-4a16-b026-5234e8bef9ec">
> 
> 

Sure, I will look out in the evening after exam

rithviknishad commented on public/logo.png:
Keep this file in the data repo instead. And update the loadOrgData script to copy it to this dir, similar to how the `logo.webp` is done.

rithviknishad reviewed (changes requested):
LGTM, just a minor change and also merge conflicts needs to be resolved.

rithviknishad commented on scripts/loadOrgData.js:
```diff
@@ -49,3 +49,16 @@ if (fs.existsSync(path.join(dataRepoPath, "config/assets"))) {
     { recursive: true },
   );
 }
+const additionalFileSrc = path.join(dataRepoPath, "config/opengraph-image.tsx");
+const additionalFileDest = path.join(
+  cwd,
+  "app/contributors/[slug]",
+  "opengraph-image.tsx",
+);
+
+if (fs.existsSync(additionalFileSrc)) {
+  console.log("Copying additional file opengraph-image.tsx...");
+  fs.copyFileSync(additionalFileSrc, additionalFileDest);
+} else {
+  console.error("File opengraph-image.tsx not found in the config folder");
+}
```
Why copy the typescript file through load org data? It's not part of the data-repo right?
